packages/flutter_tools/lib/src/runner/target_devices.dart

```diff
@@ -490,15 +524,44 @@ class TargetDevicesWithExtendedWirelessDeviceDiscovery extends TargetDevices {
         includeDevicesUnsupportedByProject: includeDevicesUnsupportedByProject,
         includeDisconnected: true,
       );
+
       if (devices.length == 1) {
         Device? matchedDevice = devices.first;
+
+        if (matchedDevice is IOSDevice && !matchedDevice.devModeEnabled! ) {
+          return _handleDevModeDisabledIosDevice(matchedDevice);
+        }
+
         if (!matchedDevice.isConnected && matchedDevice is IOSDevice) {
           matchedDevice = await _waitForIOSDeviceToConnect(matchedDevice);
         }
+
         if (matchedDevice != null && matchedDevice.isConnected) {
           return <Device>[matchedDevice];
         }
       }
+
+      if (devices.length > 1) {
+        final List<Device> disabledDevices = [];
+
+        for (final Device device in devices) {
+          if (device is IOSDevice && !device.devModeEnabled!) {
+            disabledDevices.add(device);
+
+            // otherwise, just show the error and move on
+            _logger.printStatus(
+              userMessages.flutterSpecifiedDeviceDevModeDisabled(device.name),
+            );
+          }
+        }
+
+        if (disabledDevices.length == devices.length) {
```

Comment:
I think moving this outside of the loop slightly changed the way you wanted it to work.

I believe your intended functionality was that if all matching devices found had dev mode disabled, that you printed an message for each and then listed out the devices.

With how it stand currently, it will print a message each device with dev mode turned off, regardless of if there are other devices that do have it turned on.

It's a little less efficient but I think it'd be better to 
* Use `where()` to filter `devices` to only those where `device is IOSDevice && !device.devModeEnabled`.
* Then if `disabledDevices.length == devices.length`, loop through `disabledDevices`, print message for each. 
* Also, `_printUnsupportedDevices` (within the if but not within the loop)

Reply:
i wouldn't worry about the performance - it's a list with just 5-ish items, not 5000. 

I was confused by the current code, so better to make the logic simpler.

Reply:
Thats how I wanted it to work actually - it's supposed to print the message for each dev disabled device. I think it makes sense this way because there otherwise no purpose to have a device connected dev mode is off, so imo it's better to tell the user.

Reply:
Additionally, since there can be partial matches when specifying device name/id, I feel like it makes sense to at least show why we run the device that we do(if that makes sense)
eg if I have two devices, iPhone(1) and iPhone(2), and 1 is disabled
When I do flutter run -d iPhone
I think it's less pleasant if we directly run iPhone(2) without sort showing why iPhone(1) wasn't run.